File: Application Development/latency/latency_tester.py
import ping3
import sys
from statistics import mean

def ping_server(host, count=4):
    latencies = []
    lost_packets = 0

    print(f"Pinging {host} with {count} packets...")

    for i in range(count):
        latency = ping3.ping(host)
        if latency is None:
            print(f"Packet {i+1} lost.")
            lost_packets += 1
        else:
            print(f"Packet {i+1}: {latency*1000:.2f} ms")
            latencies.append(latency)

    if latencies:
        avg_latency = mean(latencies) * 1000
        print(f"\nAverage Latency: {avg_latency:.2f} ms")
    else:
        avg_latency = None
        print("\nNo packets were successfully received.")

    packet_loss = (lost_packets / count) * 100
    print(f"Packet Loss: {packet_loss:.2f}%")

    return avg_latency, packet_loss

if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("Usage: python3 latency_tester.py <server>")
        sys.exit(1)

    host = sys.argv[1]
    ping_server(host)

File: Application Development/latency/latency_tester.js
const ping = require('ping');
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function pingServer(host, count = 4) {
    let latencies = [];
    let lostPackets = 0;

    console.log(`Pinging ${host} with ${count} packets...`);

    const promises = Array.from({ length: count }, (_, i) => {
        return ping.promise.probe(host)
            .then(res => {
                if (res.alive) {
                    console.log(`Packet ${i + 1}: ${res.time} ms`);
                    latencies.push(parseFloat(res.time));
                } else {
                    console.log(`Packet ${i + 1} lost.`);
                    lostPackets++;
                }
            });
    });

    Promise.all(promises).then(() => {
        if (latencies.length > 0) {
            const avgLatency = latencies.reduce((a, b) => a + b, 0) / latencies.length;
            console.log(`\nAverage Latency: ${avgLatency.toFixed(2)} ms`);
        } else {
            console.log("\nNo packets were successfully received.");
        }

        const packetLoss = (lostPackets / count) * 100;
        console.log(`Packet Loss: ${packetLoss.toFixed(2)}%`);

        rl.close();
    });
}

rl.question("Enter server IP or hostname: ", (host) => {
    pingServer(host);
});
